modules/reindex/src/main/java/org/elasticsearch/index/reindex/RemoteScrollableHitSource.java

```diff
@@ -0,0 +1,597 @@
+
+public class RemoteScrollableHitSource extends ScrollableHitSource {
+    private final CloseableHttpClient httpClient = HttpClients.createMinimal(); // NORELEASE replace me with the Elasticsearch client
+    private final RemoteInfo remoteInfo;
+    private final SearchRequest searchRequest;
+    Version remoteVersion;
+
+    public RemoteScrollableHitSource(ESLogger logger, BackoffPolicy backoffPolicy, ThreadPool threadPool, Runnable countSearchRetry,
+            Consumer<Throwable> fail, RemoteInfo remoteInfo, SearchRequest searchRequest) {
+        super(logger, backoffPolicy, threadPool, countSearchRetry, fail);
+        this.remoteInfo = remoteInfo;
+        this.searchRequest = searchRequest;
+        if (remoteInfo.getUsername() != null) {
+            // NORELEASE switch to the elasticsearch client
+            throw new UnsupportedOperationException("TLS is totally unsupported until we switch to the Elasticsearch client");
+        }
+    }
+
+    @Override
+    protected void startImpl(Consumer<? super InternalResponse> onResponse) {
+        threadPool.generic().execute(runnable(() -> {
+            remoteVersion = lookupRemoteVersion();
+            HttpPost post = buildStartRequest();
+            executeToInternalResponse(post, r -> {
+                if (Strings.hasLength(r.getScroll()) && r.getHits().isEmpty()) {
+                    logger.debug("First response looks like a scan response. Jumping right to the second. scroll=[{}]", r.getScroll());
+                    startNextScrollImpl(r.getScroll(), timeValueMillis(0), onResponse);
+                } else {
+                    onResponse.accept(r);
+                }
+            });
+        }));
+    }
+
+    Version lookupRemoteVersion() {
+        HttpResponse response = execute(new HttpGet(uri().toString()));
+        try {
+            try (XContentParser parser = XContentFactory.xContent(response.body).createParser(response.body)) {
+                return MAIN_ACTION_PARSER.apply(parser, () -> ParseFieldMatcher.STRICT);
+            }
+        } catch (IOException e) {
+            throw new RuntimeException("Failed to parse main action response=[" + response.body + "]");
+        }
+    }
+
+    HttpPost buildStartRequest() {
+        StringBuilder uri = uri();
+        for (String[] array : new String[][] { searchRequest.indices(), searchRequest.types() }) {
+            if (array != null && array.length > 0) {
+                uri.append(array[0]);
+                for (int i = 1; i < array.length; i++) {
+                    uri.append(',').append(array[i]);
+                }
+                uri.append('/');
+            }
+        }
+        uri.append("_search?scroll=").append(searchRequest.scroll().keepAlive());
+        uri.append("&size=").append(searchRequest.source().size());
+        if (searchRequest.source().version()) {
+            uri.append("&version");
+        }
+        if (searchRequest.source().sorts() != null) {
```

Comment:
Is `smile` the lowest overhead of our supported formats? It has been useful to be able to read the json but I'm happy to switch it. I'd love to reduce the overhead of the thing.